pkg/chunked/cache_linux.go

```diff
@@ -91,83 +101,135 @@ func getLayersCache(store storage.Store) (*layersCache, error) {
-		if err != nil && !errors.Is(err, os.ErrNotExist) {
-			return err
+	cacheFile, err := readCacheFileFromMemory(buffer)
+	if err != nil {
+		if mmapBuffer != nil {
+			unix.Munmap(mmapBuffer)
 		}
-		if clFile != nil {
-			cl, err := io.ReadAll(clFile)
-			if err != nil {
-				return fmt.Errorf("open manifest file for layer %q: %w", r.ID, err)
-			}
-			json := jsoniter.ConfigCompatibleWithStandardLibrary
-			if err := json.Unmarshal(cl, &lcd); err != nil {
-				return err
-			}
+		return false, err
+	}
+	if err := c.addLayer(layerID, cacheFile, mmapBuffer); err != nil {
+		// the mmap'ed data is not owned by the cache manager on errors
+		if mmapBuffer != nil {
+			unix.Munmap(mmapBuffer)
 		}
+		return false, err
+	}
+	return true, nil
+}
 
-		// otherwise create it from the layer TOC.
-		manifestReader, err := c.store.LayerBigData(r.ID, bigDataKey)
-		if err != nil {
+func (c *layersCache) createCacheFileFromTOC(layerID string) error {
+	clFile, err := c.store.LayerBigData(layerID, chunkedLayerDataKey)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	if clFile == nil {
+		return nil
+	}
+	cl, err := io.ReadAll(clFile)
+	if err != nil {
+		return fmt.Errorf("open manifest file: %w", err)
+	}
+	json := jsoniter.ConfigCompatibleWithStandardLibrary
+
+	var lcd chunkedLayerData
+	if err := json.Unmarshal(cl, &lcd); err != nil {
+		return err
+	}
+	manifestReader, err := c.store.LayerBigData(layerID, bigDataKey)
+	if err != nil {
+		return err
+	}
+	defer manifestReader.Close()
+
+	manifest, err := io.ReadAll(manifestReader)
+	if err != nil {
+		return fmt.Errorf("read manifest file: %w", err)
+	}
+
+	cacheFile, err := writeCache(manifest, lcd.Format, layerID, c.store)
+	if err != nil {
+		return err
+	}
+	return c.addLayer(layerID, cacheFile, nil)
+}
+
+func (c *layersCache) load() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	existingLayers := make(map[string]struct{})
+	for _, r := range c.layers {
+		// ignore the layer if it was fully loaded in memory.
+		// In this way it can be reloaded using mmap.
+		if r.mmapBuffer != nil {
+			existingLayers[r.id] = struct{}{}
+		}
+	}
+	allLayers, err := c.store.Layers()
+	if err != nil {
+		return err
+	}
+	for _, r := range allLayers {
+		if _, found := existingLayers[r.ID]; found {
 			continue
 		}
-		defer manifestReader.Close()
-
-		manifest, err := io.ReadAll(manifestReader)
+		// try to read the existing cache file
+		loaded, err := c.loadLayerCache(r.ID)
 		if err != nil {
-			return fmt.Errorf("open manifest file for layer %q: %w", r.ID, err)
+			logrus.Warningf("Error loading cache file for layer %q: %v", r.ID, err)
 		}
-
-		metadata, err := writeCache(manifest, lcd.Format, r.ID, c.store)
-		if err == nil {
-			c.addLayer(r.ID, metadata)
+		if loaded {
+			continue
 		}
-	}
-
-	var newLayers []layer
-	for _, l := range c.layers {
-		if _, found := currentLayers[l.id]; found {
-			newLayers = append(newLayers, l)
```

Comment:
It's a bit suspicious that you've removed this part (which re-sets `c.layers`). My guess is, either you assume that 
- there can be no situation when some layers are gone during a subsequent call to `getLayersCache(store)`, or
- it is not important that some (stale?) layers are not removed from the cache.

Please clarify.

Reply:
you are right the new code didn't deal correctly with stale layers.

I've changed the code to cleanup stale layers and override `c.layers` as before